### ISSUE TITLE
Incorrect IDE004 "Cast is redundant" hint

### ISSUE THREAD

**Issue by IharBury** [dotnet/roslyn#11008]
**Version Used**: 
Microsoft Visual Studio 2015 14.0.25123.0 Update 2
.NET Framework 4.6.01038

**Steps to Reproduce**:
Write code:

``` c#
static class Program
{
  static readonly long x = -(long)int.MinValue;
}
```

**Expected Behavior**:
No compiler hints

**Actual Behavior**:
IDE0004 Cast is redundant

If the cast is removed:
CS0220 The operation overflows at compile time in checked mode

**Comment by Allon-Guralnek** [comment]
I have another case where I get an incorrect: "Cast is redundant". I encountered this using Visual Studio Update 3. 

Given any enum type named `E`, the following doesn't display any warning:

```csharp
int? num = 1;
string s = ((E?)num)?.ToString();
```

But chaining another method does cause this warning to be reported:

```csharp
int? num = 1;
string s = ((E?)num)?.ToString().Replace('a', 'b');
```

The `(E?)` above will have a "Cast is redundant" warning. But removing the cast will change the value of the string `s`, so the cast is not redundant.

**Comment by IharBury** [comment]
@Allon-Guralnek, maybe it's not exactly redundant, but it doesn't make sense either. When `num` is `null`, `?.ToString()` results in `null` and `.Replace` throws `NullReferenceException`. So the correct code would use `?.Replace` or would work with non-nullable `int`.

**Comment by Allon-Guralnek** [comment]
@IharBury, are you sure? I just tried running it when `num = null` and no NullReferenceException was thrown. The value of `s` was simply `null`.

**Comment by tamlin-mike** [comment]
Amazing. Soon four years since this was reported, and still pushed back. Sweet lord...

"help wanted"? More like _"NEED HELP!" (to replace managers)_.

To put some more fuel on the fire (different code, but still IDE0004):
```
class C
{
  static uint fn1(sbyte a, sbyte b)
  {
    return (uint)((a << 8) | (int)b); // IDE0004
  }
  public uint fn2(sbyte a, sbyte b)
  {
    return (uint)((a << 8) | b); // CS0675
  }
}
```

_Please don't let those managers continue to do side-gigs for Boeing -- you could be on the next plane._

**Comment by CyrusNajmabadi** [comment]
@tamlin-mike Would you like to contribute a fix here?  I can walk you through where it would go.  It sounds like you're enthusiastic about this issue.  It would def be great to be able to take that interest and allow you to fix the things impacting you :)

LMK!
